Add several functions from `box.session` to module API

olmiik opened the pull request:
The following functions from `box.session` are added to module API:
- `box_session_id`
- `box_session_exists`
- `box_session_peer`
- `box_session_user`
- `box_session_type`
- `box_session_uid`
- `box_session_euid`

Only trivial functions were added. We have to put some thought into implementations of `box.session.su()`,  `box.session.on_connect()`,  `box.session.on_disconnect()` `box.session.on_auth()` and can add them via future pull requests.

Gerold103 commented on src/box/session.cc:
```diff
@@ -37,6 +37,8 @@
 #include "tt_static.h"
 #include "sql_stmt_cache.h"
+#include "sio.h"
+#include "box.h"
```
Why do you need box.h header?

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +377,71 @@ generic_session_sync(struct session *session)
+API_EXPORT bool
+box_session_exists(uint64_t sid)
+{
+	return (session_find(sid) != NULL);
```
You don't need the extra braces around.

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +377,71 @@ generic_session_sync(struct session *session)
+{
+	struct session *session = session_find(sid);
+	if (session == NULL)
+		session = current_session();
```
That looks wrong. If I would pass any random number not matching a session, the last thing I would expect is to get my own peer. It is also inconsistent with the Lua API, which throws an error if couldn't find the session.

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +377,71 @@ generic_session_sync(struct session *session)
+	struct session *session = session_find(sid);
+	if (session == NULL)
+		session = current_session();
+	if (session == NULL)
```
How is it possible second time? `current_session()` is not exception-safe, so if it couldn't create a session, it throws. Not returns `NULL`. Which is actually an issue existing before this patch, and because of which even `box_session_id()` can fail. Probably we must add `panic()` in case a session couldn't be created. In a separate commit of course.

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +377,71 @@ generic_session_sync(struct session *session)
+	if (sio_getpeername(fd, (struct sockaddr *)&addr, &addrlen) < 0)
+		return NULL;
+
+	return sio_strfaddr((struct sockaddr *)&addr, addrlen);
```
`sio_strfaddr()` returns a string allocated on the internal so called "static allocator". It is not documented, and is not safe to use in userspace code. I think it makes sense to make the user pass his own buffer. Just like `inet_ntop()` does.

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +377,71 @@ generic_session_sync(struct session *session)
+}
+
+API_EXPORT int
+box_session_sync()
```
For functions not having an argument we specify `void` in the arglist. It actually makes difference making the final executable file smaller.

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +377,71 @@ generic_session_sync(struct session *session)
+	return sio_strfaddr((struct sockaddr *)&addr, addrlen);
+}
+
+API_EXPORT int
```
Seems like a bug. `sync` must have type `int64_t`. And it does in some places. However `session_sync()` somewhy implicitly casts it to `int`. We must fix it separately.

Gerold103 commented on src/box/session.h:
```diff
@@ -375,6 +375,73 @@ generic_session_fd(struct session *session);
+/**
+ * Get the unique identifier (ID) for the current session.
+ *
+ * \retval session ID
```
As you probably see, there is no a single place in `session.h` which uses `\` instead of `@` for doxygen comments. Please, conform.

Gerold103 commented on src/box/session.h:
```diff
@@ -375,6 +375,73 @@ generic_session_fd(struct session *session);
+ * Get host address and port of the session peer (if it exists).
+ *
+ * \param sid session ID
+ * \retval peer's host and port on success, emtpy string on error.
```
It is not true - you return `NULL` on error.

Gerold103 commented on src/box/session.h:
```diff
@@ -375,6 +375,73 @@ generic_session_fd(struct session *session);
+/**
+ * Get value of sync integer constant used in binary protocol.
+ *
+ * \retval sync integer constant
```
What if the session is not for an iproto connection? What it returns for a locally created fiber's session?

Gerold103 commented on src/box/session.h:
```diff
@@ -375,6 +375,73 @@ generic_session_fd(struct session *session);
+
+/**
+ * Get session type: "binary", "console",
+ * "replication", "background"
```
This is not true either - we have more session types. Some of them even are documented for the Lua API: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_session/type/.

Gerold103 commented on src/box/session.h:
```diff
@@ -375,6 +375,73 @@ generic_session_fd(struct session *session);
 int64_t
 generic_session_sync(struct session *session);
 
```
Public functions must appear in `module.h`. For that you must export their declarations to this file, which is built automatically. Look for `/** \cond public */` and `/** \endcond public */` in the code. Did you test that patch in your own application? Because if `module.h` is not populated with the new functions, you simply couldn't see their declarations.

Gerold103 commented on src/box/session.h:
```diff
@@ -375,6 +375,73 @@ generic_session_fd(struct session *session);
+ * \retval effective user ID
+ */
+API_EXPORT int
+box_session_euid();
```
What does it return for sessions created before `box.cfg{}`? The ones in locally created fibers.

Gerold103 reviewed (changes requested):
Hi! Thanks for the patch! I am not against exposing these functions. But I have to ask - why do you need them in C? What is the use case?

Talking of the patch details:
* You need to add tests.
* To export a function it is not enough to add `API_EXPORT`. Take a look at `exports.h` file - you need to add the exported functions there.
* Since you change the public API, you must create a documentation request in the commit message. Take a look at the syntax [here](https://github.com/tarantool/docbot#docbot---tarantool-documentation-pipeline-bot) and examples here: `git log` and grep by `TarantoolBot`.

olmiik commented:
> Hi! Thanks for the patch! I am not against exposing these functions. But I have to ask - why do you need them in C? What is the use case?
> 
> Talking of the patch details:
> 
> * You need to add tests.
> * To export a function it is not enough to add `API_EXPORT`. Take a look at `exports.h` file - you need to add the exported functions there.
> * Since you change the public API, you must create a documentation request in the commit message. Take a look at the syntax [here](https://github.com/tarantool/docbot#docbot---tarantool-documentation-pipeline-bot) and examples here: `git log` and grep by `TarantoolBot`.

Hi. Working on this one now. Sorry for poor quality. This is just refactored copy-paste of old https://github.com/tarantool/tarantool/issues/2961 

The use-case for exposing `box.session` functions in Module API can look like this:
1) Here - https://github.com/picodata/tarantool-module, we need to "re-implement" code from `src/box/schema.lua` (creation of new space) in Rust - https://github.com/picodata/tarantool-module/blob/master/src/space.rs#L141-L155
2) Without  having `box.session.uid` in Module API, we need to call them via Lua API like this -https://github.com/picodata/tarantool-module/blob/master/src/session.rs#L14-L40. Instead of just importing `box_session_uid` via FFI (like we do here - https://github.com/picodata/tarantool-module/blob/master/src/ffi/tarantool.rs).
3) I guess, the same would be true in case of development of wrappers/SDKs in other languages like C++, etc.

olmiik commented:
@Gerold103 Fixed most of your comments.

I have removed `box_session_sync` from PR as is does not used anywhere in Tarantool code, except of tests.

For tests I rewrote the following snippets from `test/box-tap/session.test.lua`  in C and put them into `test/app-tap/module_api.c`:
- https://github.com/tarantool/tarantool/blob/master/test/box-tap/session.test.lua#L23-L28
- https://github.com/tarantool/tarantool/blob/master/test/box-tap/session.test.lua#L31-L35
- https://github.com/tarantool/tarantool/blob/master/test/box-tap/session.test.lua#L117-L118

Porting of other tests requires adding to Module API of triggers from `box.session` and `box.session.su`, which are out of scope of this PR.

Gerold103 commented:
Hi! Please, read the coding style and review guidelines to see how the commits must look and how to structure them, how to write commit messages and titles: https://github.com/tarantool/tarantool/wiki/Code-review-procedure. Also you can look at `git log` to see examples. Then ensure the commits are all atomic. And then we can look at the code, thanks.

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +376,72 @@ generic_session_sync(struct session *session)
+
+	struct sockaddr_storage addr;
+	socklen_t addrlen = sizeof(addr);
+	if (sio_getpeername(fd, (struct sockaddr *)&addr, &addrlen) < 0)
```
The same - the function does not set a diag.

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +376,72 @@ generic_session_sync(struct session *session)
+	}
+	strcpy(dst, tmp);
+
+	return tmp;
```
You still return the temporary buffer which might be wiped out any moment and is not public. I also not sure what to do with the current state of copying. The user does not have a way to learn the size of the buffer he must provide as `dst`. In addition I see `box_session_user()` returns a string not belonging to a user as well. But here at least he can do `strcpy()` from his own code.

Maybe we should rethink this one and allow to return the temporary buffer here, but warn the user that it must be copied immediately? Via the documentation and comments.

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +376,72 @@ generic_session_sync(struct session *session)
+box_session_user()
+{
+	struct user *user = user_by_id(current_session()->credentials.uid);
+	if (user)
```
For non-boolean values, please, use explicit `==` and `!=`. Here it would be `user != NULL`.

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +376,72 @@ generic_session_sync(struct session *session)
+	return session_type_strs[current_session()->type];
+}
+
+API_EXPORT int
```
UID is uint32_t, you can't return it as int.

Gerold103 commented on src/box/session.cc:
```diff
@@ -375,3 +376,72 @@ generic_session_sync(struct session *session)
+}
+
+API_EXPORT int
+box_session_euid()
```
The same.

Gerold103 commented on src/box/session.h:
```diff
@@ -375,6 +375,69 @@ generic_session_fd(struct session *session);
+ * @retval peer's host and port on success, NULL on error.
+ */
+API_EXPORT const char *
+box_session_peer(uint64_t sid, char *dst, size_t size);
```
Hm. Did you think about using struct sockaddr as an out parameter? Maybe the user does not want a string. But if he wants, he can just do his own conversion. This would help us to get rid of the buffer copying problem. Just make the function work like `getpeername()`. So it takes a session ID, `struct sockaddr *addr`, and `uint32_t *addrlen`. It returns `int`. If 0, it was a success, and the address is filled with the peer address. Addrlen is an in-out parameter. After the call it contains the number of bytes written to addr. Just see `man getpeername` for details.

Note, I intentionally didn't use `socklen_t`, because it would force us to include socket headers into module.h. But with the proposed way you can just announce the sockaddr struct in module.h like this: `struct sockaddr;`.

olmiik replied:
I moved to implementation of `box_session_peer` from old PR (the same, that you are proposing now).

Gerold103 replied:
Hm. I don't remember you using `struct sockaddr` in any of your old PRs. Could you please give a link?

Gerold103 commented on src/box/session.h:
```diff
@@ -375,6 +375,69 @@ generic_session_fd(struct session *session);
+
+/**
+ * Get the type of connection or cause of action. Can be
+ * "binary", "console", "repl", "applier" or "background"
```
Lets also add a comment, that more session types might appear in the future, and some of these might be deleted/renamed. This is a string, not a enum or an ID, so anything can happen.

Gerold103 commented on src/box/session.h:
```diff
@@ -375,6 +375,69 @@ generic_session_fd(struct session *session);
+box_session_peer(uint64_t sid, char *dst, size_t size);
+
+/**
+ * Get user's name for current session.
```
Please, add a comment that the returned string memory does not belong to the user and might be deleted on any operation on the users, or on any yield. It must be copied if the user wants to keep it.

Gerold103 commented on test/app-tap/module_api.c:
```diff
@@ -2378,6 +2452,9 @@ luaopen_module_api(lua_State *L)
 		{"tuple_validate_def", test_tuple_validate_default},
 		{"tuple_validate_fmt", test_tuple_validate_formatted},
 		{"test_key_def_dup", test_key_def_dup},
+		{"session_check_admin", session_check_admin},
```
You probably noticed, there are 38 lines above in all of which the functions have `test_` prefix. Please, conform. The test case functions must start with `test_`.

olmiik replied:
They do not have prefix "test",  because they are called here "manually" here - https://github.com/picodata/tarantool/blob/module-api-add-box-session/test/app-tap/module_api.test.lua#L256-L261

while other tests are called here "automatically" - https://github.com/picodata/tarantool/blob/module-api-add-box-session/test/app-tap/module_api.test.lua#L242-L243

The initial idea was to reproduce `app-tap/module_api.test.lua`. It has some logic between tests like creation/dropping of tests and so on.

Gerold103 commented on test/app-tap/module_api.c:
```diff
@@ -2335,6 +2335,80 @@ test_tuple_validate_formatted(lua_State *L)
+	sess_id = box_session_id();
+	char peer_str_arg[100];
+	const char *peer_str = box_session_peer(sess_id, peer_str_arg, 100);
+    bool peer_is_null = (peer_str == NULL);
```
The alignment is broken. It must contain 1 tab, not 4 spaces.

Gerold103 commented on test/app-tap/module_api.c:
```diff
@@ -2335,6 +2335,80 @@ test_tuple_validate_formatted(lua_State *L)
+	bool sess_exist = box_session_exists(sess_id);
+	assert(sess_exist);
+
+	sess_id = box_session_id();
```
But you just got the session id on the line 2341, why do you call it again?

Gerold103 commented on test/app-tap/module_api.c:
```diff
@@ -2335,6 +2335,80 @@ test_tuple_validate_formatted(lua_State *L)
+{
+	(void) va;
+	do {
+		fiber_set_cancellable(true);
```
Isn't it cancelable by default? Why do you need to call it?

olmiik replied:
See my comment below.

Gerold103 commented on test/app-tap/module_api.c:
```diff
@@ -2335,6 +2335,80 @@ test_tuple_validate_formatted(lua_State *L)
+	(void) va;
+	do {
+		fiber_set_cancellable(true);
+		fiber_sleep(0.01);
```
Why do you need a sleep? Why can't the whole function just return the session ID and exit?

olmiik replied:
The logic for launching background fiber with some work and waiting for it to exit is copied from here https://github.com/tarantool/tarantool/blob/master/test/app-tap/module_api.c#L258-L285

Gerold103 replied:
I didn't review the patch yet, will do later, but I just warn you - the fact that something is done in some other place in one way does not mean you need to blindly copy it in the new code. Old code != good code in 100% cases. It is never a justification, not in a bit. Please, revisit the comments where you justify them as 'they are copied' bearing this in mind before I will do a next review.

Gerold103 commented on test/app-tap/module_api.c:
```diff
@@ -2335,6 +2335,80 @@ test_tuple_validate_formatted(lua_State *L)
+		fiber_sleep(0.01);
+		if (fiber_is_cancelled()) {
+			int sess_id = box_session_id();
+			if (sess_id == 0) {
```
This is a really strange test. Why is 0 session ID considered a fail? Maybe better return the session ID via out parameters and in the test function check it does not equal to the current session?

olmiik replied:
This part of tests is copy of https://github.com/tarantool/tarantool/blob/master/test/box-tap/session.test.lua#L31-L35
The logic of test is fully borrowed from there.

Gerold103 commented on test/app-tap/module_api.c:
```diff
@@ -2335,6 +2335,80 @@ test_tuple_validate_formatted(lua_State *L)
 }
 
+static int
+session_check_admin(lua_State *L)
```
Why is it called check admin, if in this function nothing is related to admin?

olmiik replied:
This part of `modul_api.c` copied from https://github.com/tarantool/tarantool/blob/master/test/box-tap/session.test.lua#L21-L28. Maybe, another name is more appropriate.

Gerold103 commented on test/app-tap/module_api.test.lua:
```diff
@@ -252,7 +252,12 @@ require('tap').test("module_api", function(test)
+
     space:drop()
+
+    test:ok(module.session_check_after_cleanup(), "session_check_after_cleanup is ok")
```
- You didn't test session peer when there is actually a peer;
- You didn't test getting peer of a not existing session;
- You didn't test having a not big enough buffer for a peer name;
- You didn't test getting UID when the user is deleted (you can drop the user after doing `su` to him, AFAIR);
- You didn't test getting EUID, and that it might be different from UID;
- You didn't test getting session type.

olmiik replied:
As I mentioned in previous comments the idea was to reproduce tests from `app-tap/module_api.test.lua`. Not all of them can be reproduced, because no everything from `box.session` Lua API is added to Module API in this PR (for example, `box.session.su` is not added). All 3 tests, that you see currently are snippets from `app-tap/module_api.test.lua`.

What is you opinion on this? What tests from list above must be added?

Gerold103 replied:
Personally, I would not look at module_api.test.lua as an example. I would try to write new tests, easier to read and extend. But covering at least the same cases and maybe more. Although it does not mean you need to do it (repeat something != justification, including blindly doing what I said). Maybe you can copy logic from the module_api.test.lua test but do some minor changes to address the strange things I noticed and questions I asked in the other comments.

As for the mentioned list of tests in the comment above - I think you need to cover them all. They do not look hard to test, rather easy, but otherwise you simply have code not tested a single time anywhere. We can't merge such code (unless it is urgent and some managers are running for their yearly bonus of course, haha).

Gerold103 reviewed (commented):
Hi! Good job on the patch! It looks better now, see some new comments below.

olmiik commented on test/app-tap/module_api.c:
```diff
@@ -2335,6 +2335,80 @@ test_tuple_validate_formatted(lua_State *L)
+}
+
+static int
+session_check_after_cleanup(lua_State *L)
```
This part of tests is copy of https://github.com/tarantool/tarantool/blob/master/test/box-tap/session.test.lua#L115-L118 
`space:drop() -- tweedledum` is referred as "cleanup" and is located here in `module_api.test.lua` - https://github.com/picodata/tarantool/blob/module-api-add-box-session/test/app-tap/module_api.test.lua#L259

Again, maybe, another name for subtest is more appropriate.

kostja commented:
* Vladislav Shpilevoy ***@***.***> [21/05/06 23:33]:
> As for the mentioned list of tests in the comment above - I think you need to cover them all. They do not look hard to test, rather easy, but otherwise you simply have code not tested a single time anywhere. We can't merge such code (unless it is urgent and some managers are running for their yearly bonus of course, haha).

No one is running for a bonus here. In fact we do not strictly
need this patch any more, we worked around it. The purpose of this
patch is to begin a larger effort of rewriting most of schema.lua
in C, to make it available to all programming languages and
plug-ins like rust, without having to marshal things to Lua and
invoke Lua just to create a space.

I think if you try to rewrite  vshard internally using the new C++
connector you'll see the same problem.

Regarding your test comments, while I agree in spirit with them,
in practice I think you're being too strict. It's going to be hard
to cover all cases with the current testing infrastructure, and
Oleg is doing the best he can.  

-- 
Konstantin Osipov, Moscow, Russia

Gerold103 commented:
> No one is running for a bonus here

I am not talking about you and not about this PR. Only about some decisions I didn't like about other tickets and patches on which our team worked recently and which were pushed to master with bugs even though I was very much against them as the bugs were severe and well known.

Yes, Oleg is doing good, and I am trying to help giving more detailed comments than I usually do, and I am already far from being as strict as I am for internal patches.

Talking of the tests cases - they are trivial to test. Just trivial. It would take 1-2h to cover them all, and it would be faster than try to prove that they are not needed or are too complex.

Talking of vshard - it is in progress, AFAIK. But no complains so far about missing APIs from the author of the rewrite.

Totktonada commented:
@omikee What is a status of the PR? Are all Vlad's comments fixed now?
